### PR TITLE
Added item description

### PR DESCRIPTION
**Pull request by quasipickle**
I'd like to retrieve & display the item description, not the item content.  So, I added findItemDescription().

### CODE REVIEW

**Comment by mkresin** [comment]
Have you considered using ```getTag()``` as illustrated in the [feed-parsing documentation](https://github.com/fguillot/picoFeed/blob/master/docs/feed-parsing.markdown#get-raw-xml-tagsattributes-or-non-standard-tags-for-items)?

Your commit [breaks code](https://travis-ci.org/fguillot/picoFeed/jobs/58631989). Please use the supplied tests and fix it.

@Raydiation Do you see a use case where explicit the description is needed?

**Comment by Raydiation** [comment]
The default implementation first looks for namespaced encoded elements and uses their value, then falls back to the description. There is a content module for RSS that lets you differentiate between text and HTML, encoded provides HTML, thats why encoded is preferred. Also popular feeds like golem.de use this module http://golem.de.dynamic.feedsportal.com/pf/578068/http://rss.golem.de/rss.php?feed=RSS1.0

The implementation is obviously a bit hacky, we would have to validate the namespace URL and not just accept any namespace that defines an encoded element, but well, it works for now. Bonus Kudos if this gets fixed

So far I dont see what this PR actually tries to achieve.

**Comment by mkresin** [comment]
One more thing. Let me explain the function findItemContent() you have copied and changed.

The purpose of the function is to fetch the content and fallback to the description if the content is empty or doesn't exist.

You don't need that logic for the description. You only need to check if the description tag exists and return either its content or an empty string.

I still can not imagine why you explicitly want to have the (short) description if you could have the full content. Would you please enlighten me!

**Comment by Raydiation** [comment]
@mkresin there is no content element, RSS defines the description element as the content element.

**Comment by Raydiation** [comment]
I guess this is probably a misunderstanding because of how the methods are named

**Comment by mkresin** [comment]
@Raydiation Oh come on. I know it wasn't a technical exact description. My reply was written to be easy understandable and not to cite the specs ;-).

**Comment by quasipickle** [comment]
**What I'm trying to achieve**
The feeds I'm reading are generated from Wordpress.  They have both a `description` and `content:encoded` element for each item.  I only want to display the content inside the `description` element.  This is for a staff portal, where the RSS feed content is only supposed to take up a small part of the screen.

Maybe this is me just misunderstanding what the library is intended for.  I'm using this as a replacement for SimplePie, not a roll-my-own type of Feedly.

Anyway, I've tweaked Rss20.php, updated Atom.php (which I missed the first time), and the tests are passing.

**Comment by mkresin** [comment]
Your code has still issues. I will comment the single lines in a few minutes.

Using the description element doesn't guaranty to have only a short summary. The description can contain full articles. It depends on the feed author how much text is provided within the description element. In your case it depends on the wordpress settings.

You solution would be, to display only the first n characters of the item content by using a function like https://github.com/miniflux/miniflux/blob/master/lib/helpers.php#L114 (take care of https://github.com/miniflux/miniflux/pull/361)

File: lib/PicoFeed/Parser/Atom.php [hunk 1]
```diff
@@ -205,6 +205,22 @@ public function findItemContent(SimpleXMLElement $entry, Item $item)
...
+     * @param  SimpleXMLElement   $entry   Feed item
+     * @param  \PicoFeed\Parser\Item     $item    Item object
+     */
+    public function findItemDescription(SimpleXMLElement $entry, Item $item)
+    {
+        if(isset($entry->summary) && ! empty($entry->summary)){
+            return (string) $entry->summary;
+        }
+
+        return '';
```
Comment (mkresin) [inline lib/PicoFeed/Parser/Atom.php]: The return is wrong here. The summary should be stored in $item->description.

Even without testing it, I doubt that the condition with isset() and empty() is needed here.

File: lib/PicoFeed/Parser/Rss20.php [hunk 2]
```diff
@@ -208,15 +208,28 @@ public function findItemAuthor(SimpleXMLElement $xml, SimpleXMLElement $entry, I
     public function findItemContent(SimpleXMLElement $entry, Item $item)
     {
         $content = XmlParser::getNamespaceValue($entry, $this->namespaces, 'encoded');
-
+
```
Comment (mkresin) [inline lib/PicoFeed/Parser/Rss20.php]: Unrelated code change

File: lib/PicoFeed/Parser/Rss20.php [hunk 3]
```diff
@@ -208,15 +208,28 @@ public function findItemAuthor(SimpleXMLElement $xml, SimpleXMLElement $entry, I
...
     /**
+     * Find the item description
+     *
+     * @access public
+     * @param  SimpleXMLElement          $entry   Feed item
+     * @param  \PicoFeed\Parser\Item     $item    Item object
+     */
+    public function findItemDescription(SimpleXMLElement $entry, Item $item)
+    {
+        $description = XmlParser::getNamespaceValue($entry, $this->namespaces, 'description');
```
Comment (mkresin) [inline lib/PicoFeed/Parser/Rss20.php]: $description isn't used within the function. It's either a copy/paste leftover or a bug

File: lib/PicoFeed/Parser/Rss20.php [hunk 4]
```diff
@@ -208,15 +208,28 @@ public function findItemAuthor(SimpleXMLElement $xml, SimpleXMLElement $entry, I
...
+        $item->description = trim((string) $entry->description);
```
Comment (mkresin) [inline lib/PicoFeed/Parser/Rss20.php]: IMHO it's not the libraries job to trim the string. You application should check the description for an empty string.

File: lib/PicoFeed/Parser/Atom.php (hunk 1 above)
Comment (quasipickle) [inline lib/PicoFeed/Parser/Atom.php]: Yep, return is wrong.  Fixing that.

The `isset()...` condition is used in `getContent()` as well, so I figured it was needed.

File: lib/PicoFeed/Parser/Rss20.php (hunk 2 above)
Comment (quasipickle) [inline lib/PicoFeed/Parser/Rss20.php]: It is unrelated to the purpose of my PR the current master branch has those spaces in and I took them out in the process of my coding.  This commit was just to restore the spaces back to the original.

File: lib/PicoFeed/Parser/Rss20.php (hunk 3 above)
Comment (quasipickle) [inline lib/PicoFeed/Parser/Rss20.php]: Bug.

**Comment by Raydiation** [comment]
I think since this method is clearly designed to only work on rss and not atom, this should simply be done using the already implemented getTag method.

Picofeeds primary purpose is feed readers so having to use a more generic method seems fine

**Comment by Raydiation** [comment]
Also because op seems to have very special edge case needs

File: lib/PicoFeed/Parser/Atom.php (hunk 1 above)
Comment (mkresin) [inline lib/PicoFeed/Parser/Atom.php]: Isset is used to check if the content element exists and prefer it in advance of summary in that case. Yeah, similar to https://github.com/fguillot/picoFeed/pull/133#issuecomment-93510853.

One polite meant advice: Stop copy'n'paste code without understanding it purpose!

File: lib/PicoFeed/Parser/Rss20.php [hunk 5]
```diff
@@ -212,11 +212,24 @@ public function findItemContent(SimpleXMLElement $entry, Item $item)
         if (trim($content) === '' && $entry->description->count() > 0) {
             $content = (string) $entry->description;
         }
-
+
```
Comment (mkresin) [inline lib/PicoFeed/Parser/Rss20.php]: another unrelated change.

**Comment by mkresin** [comment]
I agree with @Raydiation. Even with this changeset, the op will not reliable achieve his goal as pointed out in https://github.com/fguillot/picoFeed/pull/133#issuecomment-93579503

File: lib/PicoFeed/Parser/Atom.php [hunk 6]
```diff
@@ -205,6 +205,23 @@ public function findItemContent(SimpleXMLElement $entry, Item $item)
     }
 
     /**
+     * Find the item summary/description
+     *
+     * @access public
+     * @param  SimpleXMLElement   $entry   Feed item
+     * @param  \PicoFeed\Parser\Item     $item    Item object
+     */
+    public function findItemDescription(SimpleXMLElement $entry, Item $item)
```
Comment (mkresin) [inline lib/PicoFeed/Parser/Atom.php]: It has still the unneeded isset and empty. All the function need is $item->description = (string) $entry->summary;

**Comment by quasipickle** [comment]
I realize this won't *guarantee* a short description.  But the feeds I'm reading are known entities, so I can safely access the `description` specifically.  I can't fathom that I'm the only one this situation would apply to.

**Comment by Raydiation** [comment]
Try **$item->getTag('description')** ;)

**Comment by quasipickle** [comment]
Yeah - that works.

**Comment by mkresin** [comment]
@quasipickle 

Are you kidding me

That's exactly what I told you to use in my first reply. Even with a link to the documentation. But yeah, it's easier to waste other peoples time than reading the docs.

**Comment by quasipickle** [comment]
@mkresin I had written a response to your initial suggestion to use getTag() - I must have changed it before I actually saved it though.  My apologies for not ensuring I responded.

I did try using getTag() after you suggested it, and it of course worked.  But one could use getTag() for any item property, and yet other shortcut methods exist.  I'm merely trying to add another shortcut method to make this library more robust.

I'm curious how I'm "doing it wrong", by merely retrieving the description of an item.  I haven't explained what validation or checking I'm doing outside this library, because it's not relevant.  Nor is doing an out-of-bounds check relevant - as I'm just pulling the contents of an xml element - SimpleXML takes care of out-of-bounds.

All I was trying to do with this PR was add a simple method I found useful.  If you guys were going to be so pissy about it, you should have just rejected the PR outright.

**Comment by mkresin** [comment]
> I'm curious how I'm "doing it wrong", by merely retrieving the description of an item.

You general approach is bogus. You wrote: "... where the RSS feed content is only supposed to take up a small part of the screen". So your expectation is, that the ```description``` tag includes only a short summary. Even if the source is a well known entity, you should limit the amount of characters that are displayed, to make sure that only a small part of the screen is used. That is what I meant with the out of bounds check.

And as soon as you do it that way, you can use the "normal" ```getContent()``` method and there is no need to rely on the description tag. Especially since your constrain is not to get specifically the content of the description element. Or at least you never mentioned it.

> If you guys were going to be so pissy about it, you should have just rejected the PR outright.

As far as I can see, I'm the only one who is pissed. Feel free to blame me, but do not generalize it to all contributors.

**Comment by Raydiation** [comment]
Meh, can we please keep this friendly? The problem was probably that the benefit of the patch was not obvious in the beginning so everyone was confused. Props for the contribution, turned out to already be possible :) is this documented BTW?

**Comment by mkresin** [comment]
https://github.com/fguillot/picoFeed/blob/master/docs/feed-parsing.markdown#get-raw-xml-tagsattributes-or-non-standard-tags-for-items

**Comment by quasipickle** [comment]
> That is what I meant with the out of bounds check.

Gotcha.

> As far as I can see, I'm the only one who is pissed. Feel free to blame me, but do not generalize it to all contributors.

You're right - my apologies to everyone else.

I *did* see the documentation for getTag(), but since the `description` is not non-standard, nor a tag attribute, I didn't think it applied.  Admittedly I should have investigated further.

@mkresin - Despite your (our) emotional outburst, I did really appreciate your efforts commenting the code.
